### PR TITLE
feat(Partitions): display partitions for topic without consumers

### PR DESCRIPTION
**Pull request by artemmufazalov**
- Replace `getNodes` with `getNodesList` for better performance
- Use selectors for `consumers` and `nodesMap`
- Remove `PartitionsWrapper` (data fetch logic changed, it no longer fits)
- Separate `PartitionsControls` component
- Use `hiddenColumns` instead of `selectedColumns` list, since it better fits for table with different set of columns for different entities
- Enable consumer select to be without consumer. Fetch partitions data from `/describe_topic` for this case

### CODE REVIEW

File: src/types/store/consumer.ts [hunk 1]
```diff
@@ -11,32 +11,32 @@ import type {DescribeConsumerResult} from '../api/consumer';
...
 
 // All fields should be present though they could be undefined
-export interface IPreparedPartitionData {
+export interface PreparedPartitionData {
     partitionId: string;
     storeSize: string;
 
     writeSpeed: IProcessSpeedStats;
-    readSpeed: IProcessSpeedStats;
+    readSpeed: IProcessSpeedStats | undefined;
```
Comment (artemmufazalov) [inline src/types/store/consumer.ts:19]: Fields, that could be `undefined` are empty for partitions without consumers

File: src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    const handlePartitionIdSearchChange = (value: string) => {
+        setPartitionIdSearchValue(value);
+    };
+
+    const handleGeneralSearchChange = (value: string) => {
+        setGeneralSearchValue(value);
+    };
+
+    const hadleTableColumnsSetupChange = (value: TableColumnSetupItem[]) => {
```
Comment (artemmufazalov) [inline src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx:116]: Hidden columns are used now, because set of columns differs for partitions with consumers and partitions without consumers

File: src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        const result = [...hiddenColumns];
```
Comment (artemmufazalov) [inline src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx:117]: Hidden columns are used now, because set of columns differs for partitions with consumers and partitions without consumers

File: src/containers/Tenant/Diagnostics/Partitions/utils/index.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import type {PreparedPartitionData} from '../../../../../types/store/consumer';
+import type {NodesMap} from '../../../../../types/store/nodesList';
+import type {PreparedPartitionDataWithHosts} from './types';
+
+export const addHostToPartitions = (
```
Comment (artemmufazalov) [inline src/containers/Tenant/Diagnostics/Partitions/utils/index.ts:5]: It was moved from `Partitions` to reduce component size

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx [hunk 5]
```diff
@@ -43,215 +40,162 @@ export const b = block('ydb-diagnostics-partitions');
...
-
+export const Partitions = ({path}: PartitionsProps) => {
     const dispatch = useDispatch();
 
-    const [generalSearchValue, setGeneralSearchValue] = useState('');
-    const [partitionIdSearchValue, setPartitionIdSearchValue] = useState('');
-
+    // Manual path control to ensure it updates with state reset
+    // so no request with wrong params will be sent
     const [componentCurrentPath, setComponentCurrentPath] = useState(path);
```
Comment (Raubzeug) [inline src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:50]: When new `path` arrives, it triggers component's re-render. What do we need to ensure?

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx [hunk 6]
```diff
@@ -43,215 +40,162 @@ export const b = block('ydb-diagnostics-partitions');
...
-                    items={columnsToSelect}
-                    showStatus
-                    onUpdate={hadleTableColumnsSetupChange}
-                    className={b('table-settings')}
-                />
-            </div>
+            <PartitionsControls
+                consumers={consumers}
+                selectedConsumer={selectedConsumer}
+                handleSelectedConsumerChange={handleSelectedConsumerChange}
```
Comment (Raubzeug) [inline src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:189]: Just to reduce ambiguity - props should be called `onSomething`, its value - `handleSomething`

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx [hunk 7]
```diff
@@ -43,215 +40,162 @@ export const b = block('ydb-diagnostics-partitions');
...
-                title: PARTITIONS_COLUMNS_TITILES[id],
-                selected: Boolean(selectedColumns?.includes(id)),
-                id: id,
-                required: id === PARTITIONS_COLUMNS_IDS.PARTITION_ID,
-            };
-        });
-    }, [selectedColumns]);
+    }, [dispatch, topicWasLoaded, selectedConsumer, consumers]);
 
     const columnsToShow = useMemo(() => {
```
Comment (Raubzeug) [inline src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:146]: This may be moved into custom hook `useVisibleColumns = (initialColumns, hiddenColumns) => {return [visibleColumns, setColumns]}`
`hiddenColumns` - may be even key for LS, if it is common to store hidden columns for different tables in LS.

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx [hunk 8]
```diff
@@ -43,215 +40,162 @@ export const b = block('ydb-diagnostics-partitions');
...
+        dispatch(setTopicDataWasNotLoaded());
+
+        dispatch(getNodesList());
+        dispatch(getTopic(path));
+
         setComponentCurrentPath(path);
     }, [dispatch, path]);
 
-    const fetchConsumerData = useCallback(
+    // Data, its source and presentation is different for partitions with and without consumers
```
Comment (Raubzeug) [inline src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:99]: is it possible to split logic into two custom hooks - handle data with and without consumers?

File: src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+                filterable={consumers && consumers.length > 5}
+                disabled={selectIsDisabled || !consumers || !consumers.length}
+                open={consumerSelectOpen}
+                onOpenChange={setConsumerSelectOpen}
+                // Although only one value could be selected
+                // Multiple type Select is used
+                // The reason - we need Select to be able to work with no value
+                // And it is easier to make multiple Select close on value change
+                // Than to enable single select to work with empty values
+                multiple
```
Comment (Raubzeug) [inline src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx:151]: For me it is not the best way to do it, it looks like a crutch. The better choice is to make a wrapper for `Select` that will allow to undo selection (the common way is - when selected item in Select popup is clicked again, selections cancels).

File: src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+                readerName,
+                readSessionId,
+                partitionNodeId,
+                connectionNodeId,
+                partitionHost,
+                connectionHost,
+            } = partition;
+
+            const isPartitionIdMatch = partitionIdRe.test(partitionId);
+            const isOtherValuesMatch =
```
Comment (Raubzeug) [inline src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx:65]:
```
const otherValues = [readerName, readSessionId, ...].filter(Boolean).map(String)`
const isOtherValuesMatch = otherValues.some(...)
```

File: src/store/reducers/topic.ts [hunk 11]
```diff
@@ -142,4 +151,53 @@ export const selectPreparedConsumersData: Selector<
...
+            consumerReadIdleTime: undefined,
+
+            uncommitedMessages: undefined,
+            unreadMessages: undefined,
+
+            startOffset: start_offset,
+            endOffset: end_offset,
+            commitedOffset: undefined,
+
+            readSessionId: undefined,
```
Comment (Raubzeug) [inline src/store/reducers/topic.ts:194]: Could you please clarify if these properties may be not undefined?

**Review by Raubzeug — CHANGES_REQUESTED** [review]

File: src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx (hunk 9 above)
Comment (artemmufazalov) [reply src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx:151]: Single select with ability to undo selection looks like a crutch as well, but it more difficult in terms of realisation. That's why I wrote this comment. Multiple Select crutch is much easier

File: src/store/reducers/topic.ts (hunk 11 above)
Comment (artemmufazalov) [reply src/store/reducers/topic.ts:194]: They are always `undefined` for partitions from `/describe_topic` endpoint. And always defined for `/describe_consumer`

File: src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx (hunk 9 above)
Comment (artemmufazalov) [reply src/containers/Tenant/Diagnostics/Partitions/PartitionsControls/PartitionsControls.tsx:158]: https://github.com/ydb-platform/ydb-embedded-ui/issues/386

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx (hunk 8 above)
Comment (artemmufazalov) [reply src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:99]: As for it won’t make it easier, as there is a problem - there still should be some logic that will determine what entity we are dealing with

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx (hunk 7 above)
Comment (artemmufazalov) [reply src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:143]: I’m not sure that there is much use in moving columns into a custom hook. However, I created a hook `useSetting` to simplify interaction with stored settings

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx (hunk 5 above)
Comment (artemmufazalov) [reply src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:49]: No, this way it won’t render correctly on autorefresh. Also first render could happen with already loaded topic data (when navigating from info or consumers tabs)

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx (hunk 8 above)
Comment (Raubzeug) [reply src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:99]: Yes, but now this logic is splitted between useEffect and fetchData. My point was that you pass `selectedConsumer` into the hook and it returns `fetchData, columns, columnsIds` to isolate logic connected with difference you mentioned.

File: src/utils/hooks/useSetting.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+import {useTypedSelector} from './useTypedSelector';
+
+export const useSetting = <T>(key: string, defaultValue: T): [T, (value: T) => void] => {
+    const dispatch = useDispatch();
+
+    const settingValue: T = useTypedSelector(
+        (state) => getParsedSettingValue(state, key) || defaultValue,
+    );
+
+    const setValue = useCallback(
```
Comment (Raubzeug) [inline src/utils/hooks/useSetting.ts:15]: why did you decide to use `useCallback` here?

Comment (artemmufazalov) [reply src/utils/hooks/useSetting.ts:15]: I decided to add `useCallback` to ensure, that return values are properly memoized and could be safely used as dependencies in other hooks

File: src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx [hunk 13]
```diff
@@ -152,25 +130,26 @@ export const Partitions = ({path}: PartitionsProps) => {
...
+    }
 
     const getContent = () => {
-        if (loading) {
+        if (partitionsLoading && !partitionsWasLoaded) {
             return <TableSkeleton className={b('loader')} />;
         }
-        if (error) {
-            return <ResponseError error={error} />;
+        if (partitionsError) {
```
Comment (Raubzeug) [inline src/containers/Tenant/Diagnostics/Partitions/Partitions.tsx:145]: Why did you move this error to `getContent`?

**Review by Raubzeug — CHANGES_REQUESTED** [review]

**Review by Raubzeug — APPROVED** [review]

**Review by Raubzeug — APPROVED** [review]
